src/colmap/estimators/covariance.cc

```diff
@@ -0,0 +1,582 @@
+  point_blocks_.clear();
+  num_params_points_ = 0;
+  for (const double* block : point_blocks) {
+    THROW_CHECK(
+        problem_->HasParameterBlock(block) &&
+        !problem_->IsParameterBlockConstant(const_cast<double*>(block)));
+    int num_params_point = ParameterBlockTangentSize(problem_, block);
+    point_blocks_.push_back(block);
+    num_params_points_ += num_params_point;
+  }
+
+  // Parse parameter blocks for poses
+  SetPoseBlocks(pose_blocks);
+}
+
+void BundleAdjustmentCovarianceEstimatorBase::SetPoseBlocks(
+    const std::vector<const double*>& pose_blocks) {
+  // Reset block indexing map
+  map_block_to_index_.clear();
+
+  // Parse parameter blocks for poses
+  pose_blocks_.clear();
+  num_params_poses_ = 0;
+  for (const double* block : pose_blocks) {
+    THROW_CHECK(
+        problem_->HasParameterBlock(block) &&
+        !problem_->IsParameterBlockConstant(const_cast<double*>(block)));
+
+    int num_params_block = ParameterBlockTangentSize(problem_, block);
+    pose_blocks_.push_back(block);
+    map_block_to_index_.emplace(block, num_params_poses_);
+    num_params_poses_ += num_params_block;
+  }
+
+  // Parse parameter blocks for other variables
+  SetUpOtherVariablesBlocks();
+}
+
+void BundleAdjustmentCovarianceEstimatorBase::SetUpOtherVariablesBlocks() {
+  // Construct a set for excluding pose and point parameter blocks
+  std::set<const double*> pose_and_point_parameter_blocks;
+  for (const double* block : pose_blocks_)
+    pose_and_point_parameter_blocks.insert(block);
+  for (const double* block : point_blocks_)
+    pose_and_point_parameter_blocks.insert(block);
+
+  // Parse parameter blocks for other variables
+  other_variables_blocks_.clear();
+  num_params_other_variables_ = 0;
+  std::vector<double*> all_parameter_blocks;
+  problem_->GetParameterBlocks(&all_parameter_blocks);
+  for (double* block : all_parameter_blocks) {
+    if (problem_->IsParameterBlockConstant(block)) continue;
+    // check if the current parameter block is in either the pose or point
+    // parameter blocks
+    if (pose_and_point_parameter_blocks.find(block) !=
+        pose_and_point_parameter_blocks.end()) {
+      continue;
+    }
+    other_variables_blocks_.push_back(block);
+    map_block_to_index_.emplace(
+        block, num_params_poses_ + num_params_other_variables_);
+    int num_params_block = ParameterBlockTangentSize(problem_, block);
+    num_params_other_variables_ += num_params_block;
+  }
+}
+
+bool BundleAdjustmentCovarianceEstimatorBase::HasReconstruction() const {
+  return reconstruction_ != nullptr;
+}
+
+bool BundleAdjustmentCovarianceEstimatorBase::HasBlock(
+    const double* params) const {
+  return map_block_to_index_.find(params) != map_block_to_index_.end();
+}
+
+bool BundleAdjustmentCovarianceEstimatorBase::HasPoseBlock(
+    const double* params) const {
+  if (!HasBlock(params)) return false;
+  return map_block_to_index_.at(params) < num_params_poses_;
```

Comment:
Could avoid the redundant lookup and just do:
```
const auto it = map_block_to_index_.find(params);
if (it == map_block_to_index_.end()) {
  return false;
}
return it->second;
```

Reply:
Thanks. This is fixed now.